theme-kit/router.ts

```diff
@@ -0,0 +1,147 @@
+/* eslint-disable @typescript-eslint/no-use-before-define,@typescript-eslint/naming-convention */
+
+import type { Culture } from '@prezly/sdk';
+import { notFound } from 'next/navigation';
+import type { ReactElement, ReactNode } from 'react';
+import UrlPattern from 'url-pattern';
+
+import { assertServerEnv } from '@/utils';
+
+import { api } from './api';
+
+export function route<
+    Pattern extends `/${string}`,
+    Match extends ExtractPathParams<Pattern>,
+    Props = Match,
+>(pattern: Pattern, loader: Loader<Match, Props>) {
+    const urlPattern = buildUrlPattern(pattern);
+
+    return {
+        async match(path: string, getDefaultLocale, isEnabledLocale) {
+            const matched = urlPattern.match(path);
+
+            if (!matched) {
+                return undefined;
+            }
+
+            if (matched.locale && !(await isEnabledLocale(matched.locale))) {
+                return undefined;
+            }
+
+            if (!matched.locale) {
+                matched.locale = await getDefaultLocale();
+            }
+
+            const { match: pageMatch, resolveLocale, Layout, default: Page } = await loader();
+
+            const pageMatched = pageMatch ? await pageMatch(matched) : matched;
+
+            if (!pageMatched) {
+                return undefined;
+            }
+
+            if (resolveLocale) {
+                return {
+                    match: {
+                        ...pageMatched,
+                        locale: await resolveLocale(pageMatched),
+                    },
+                    Layout,
+                    Page,
+                };
+            }
+
+            return {
+                match: { ...pageMatched, locale: matched.locale ?? (await getDefaultLocale()) },
```

Comment:
This duplicates the logic on line 31

Reply:
Hmm.. :thinking:  Not really. As `pageMatched` is a different object compared to `matched`. 

But I see why you're confused, it's (too) tricky. I'll think on how it can be restructured. :+1: